### PR TITLE
refactor: export component animations for reuse

### LINKED ISSUE #8904: Animations should be exported for re-use
#### Bug, feature request, or proposal:
Feature request

#### What is the expected behavior?
Animations to be exported

#### What is the current behavior?
Animations are not exported

#### What is the use-case or motivation for changing an existing behavior?
Building custom material design components that match existing components without duplicating code. For example I would like to create sort component that would allow me to sort table on mobile devices when not every column is visible and the space is limited. Right now I'm forced to copy animations from https://github.com/angular/material2/blob/b4a83ab49a1c3d27acfdeb2ece8e21f755a32606/src/lib/sort/sort-header.ts#L69-L104 to look the same as `MatSortHeader`. 

With exported animations I could reuse that code.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
5.0.0

#### Is there anything else we should know?
Some animations *are* exported so there is an inconsistency.

https://github.com/angular/material2/blob/b4a83ab49a1c3d27acfdeb2ece8e21f755a32606/src/lib/select/select-animations.ts#L34-L69

### PR DESCRIPTION
**Pull request by crisbeto**
Exports all of the component animations so they can be reused by consumers.

Fixes #8904.

### CODE REVIEW

**Comment by josephperrott** [comment]
If we are exporting them all publically perhaps we should "namespace" them?  We could put all of the definitions inside of respective objects for animations.  I worry about polluting the exported namespace.

For example: 
```ts
export const ExpansionPanelAnimations = {
  indicatorRotate: AnimationTriggerMetadata = trigger('indicatorRotate', [
    state('collapsed', style({transform: 'rotate(0deg)'})),
    state('expanded', style({transform: 'rotate(180deg)'})),
    transition('expanded <=> collapsed',
               animate(EXPANSION_PANEL_ANIMATION_TIMING)),
  ]),
  expansionHeaderHeight: AnimationTriggerMetadata = trigger('expansionHeight', [
    state('collapsed', style({
      height: '{{collapsedHeight}}',
    }), {
      params: {collapsedHeight: '48px'},
    }),
    state('expanded', style({
      height: '{{expandedHeight}}'
    }), {
      params: {expandedHeight: '64px'}
    }),
    transition('expanded <=> collapsed', 
               animate(EXPANSION_PANEL_ANIMATION_TIMING)),
  ]),
  bodyExpansion: AnimationTriggerMetadata = trigger('bodyExpansion', [
    state('collapsed', style({height: '0px', visibility: 'hidden'})),
    state('expanded', style({height: '*', visibility: 'visible'})),
    transition('expanded <=> collapsed', 
               animate(EXPANSION_PANEL_ANIMATION_TIMING)),
  ])
}
```

**Comment by crisbeto** [comment]
Good point @josephperrott, I was hoping that this would come up during the review since I'm not a fan of just exporting them as plain variables either. Switch over to namespaces.

File: src/lib/dialog/dialog-animations.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+  style,
+  transition,
+  trigger,
+  AnimationTriggerMetadata,
+} from '@angular/animations';
+
+/**
+ * Animations used by MatDialog.
+ */
+export namespace MatDialogAnimations {
```
Comment (jelbourn) [inline src/lib/dialog/dialog-animations.ts]: We can't use typescript namespaces; they're forbidden inside google (has to do with interop with Closure-style javascript)

Comment (crisbeto) [reply src/lib/dialog/dialog-animations.ts]: In this case what would be the proper way to namespace them? Should we have a `MatDialogAnimations` class with static properties or just an object with some properties on it? The static class has the advantage of allowing us to make the properties readonly.

Comment (jelbourn) [reply src/lib/dialog/dialog-animations.ts]: I would just do an object that we can use typescript to make immutable

**Comment by crisbeto** [comment]
Refactored it based on our discussion @jelbourn. Note that I'm not a fan of having to add the typings to all of the constants, but it seems like the simplest way of ensuring that the properties are immutable. Initially I was going for something like `{readonly [key: string]: AnimationTriggerMetadata}`, but that really only prevents people from adding new keys, it doesn't break if they try to override the existing ones.

File: src/lib/dialog/dialog-animations.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+/**
+ * Animations used by MatDialog.
+ */
+export const matDialogAnimations: {
+  readonly slideDialog: AnimationTriggerMetadata;
+} = {
+  /**
+   * Animation that slides the dialog in and out of view and fades the opacity.
+   */
```
Comment (jelbourn) [inline src/lib/dialog/dialog-animations.ts]: nit: can condense these comments to a single line

**Comment by jelbourn** [comment]
Changing this to minor release since it adds new APIs

**Pull request merged**